src/libraries/System.Security.Cryptography.Cose/src/System/Security/Cryptography/Cose/CoseSign1Message.cs

```diff
@@ -17,64 +18,108 @@ internal CoseSign1Message(CoseHeaderMap protectedHeader, CoseHeaderMap unprotect
+            => SignCore(content, key, hashAlgorithm, GetKeyType(key), protectedHeaders, unprotectedHeaders, isDetached);
+
+        [UnsupportedOSPlatform("browser")]
+        internal static byte[] SignCore(ReadOnlySpan<byte> content, AsymmetricAlgorithm key, HashAlgorithmName hashAlgorithm, KeyType keyType, CoseHeaderMap? protectedHeaders, CoseHeaderMap? unprotectedHeaders, bool isDetached)
         {
-            KeyType keyType = key switch
+            ValidateBeforeSign(protectedHeaders, unprotectedHeaders, keyType, hashAlgorithm, out int? algHeaderValueToSlip);
+
+            int expectedSize = ComputeEncodedSize(protectedHeaders, unprotectedHeaders, algHeaderValueToSlip, content.Length, isDetached, key.KeySize, keyType);
+            byte[] rentedBuffer = ArrayPool<byte>.Shared.Rent(expectedSize);
+
+            try
             {
-                ECDsa => KeyType.ECDsa,
-                RSA => KeyType.RSA,
-                _ => throw new CryptographicException(SR.Format(SR.Sign1SignUnsupportedKey, key.GetType()))
-            };
+                int bytesWritten = CreateCoseSign1Message(content, rentedBuffer, key, hashAlgorithm, protectedHeaders, unprotectedHeaders, isDetached, algHeaderValueToSlip, keyType);
+                Debug.Assert(expectedSize == bytesWritten);
 
-            ThrowIfDuplicateLabels(protectedHeaders, unprotectedHeaders);
+                return rentedBuffer.AsSpan(0, bytesWritten).ToArray();
+            }
+            finally
+            {
+                ArrayPool<byte>.Shared.Return(rentedBuffer, clearArray: true);
+            }
+        }
 
-            int? algHeaderValueToSlip = ValidateOrSlipAlgorithmHeader(protectedHeaders, unprotectedHeaders, keyType, hashAlgorithm);
+        [UnsupportedOSPlatform("browser")]
+        public static bool TrySign(
+            ReadOnlySpan<byte> content, Span<byte> destination,
+            AsymmetricAlgorithm key!!, HashAlgorithmName hashAlgorithm, out int bytesWritten,
+            CoseHeaderMap? protectedHeaders = null, CoseHeaderMap? unprotectedHeaders = null, bool isDetached = false)
+        {
+            KeyType keyType = GetKeyType(key);
+            ValidateBeforeSign(protectedHeaders, unprotectedHeaders, keyType, hashAlgorithm, out int? algHeaderValueToSlip);
+
+            int expectedSize = ComputeEncodedSize(protectedHeaders, unprotectedHeaders, algHeaderValueToSlip, content.Length, isDetached, key.KeySize, keyType);
+            if (expectedSize > destination.Length)
+            {
+                bytesWritten = 0;
+                return false;
+            }
 
-            byte[] encodedProtetedHeaders = CoseHeaderMap.Encode(protectedHeaders, mustReturnEmptyBstrIfEmpty: true, algHeaderValueToSlip);
-            byte[] toBeSigned = CreateToBeSigned(SigStructureCoxtextSign1, encodedProtetedHeaders, content);
+            byte[] rentedBuffer = ArrayPool<byte>.Shared.Rent(expectedSize);
```

Comment:
"yep" sounds kinda dangerous, what if there's an exception? should we try-catch and clear `destination` in case something goes wrong?

Reply:
The paradigm for span-writing methods is "assume the contents are toast if an exception was thrown".  Clearing on exception is something we do in certain crypto cases, such as data tampering was detected over partially decrypted data, because we really don't want anyone to depend on "oh, I can use the partially decrypted data".  If it's just junk state, eh.